### DISCUSSION TITLE
Problems with Date casting when using model data binding

### DISCUSSION THREAD

**Discussion by TiagoSilvaPereira** [livewire/livewire#3184] [7. Bugs (archived)] [answered]
Hi there,

I have a Laravel application in that all dates are being cast to Carbon instances. 

```php
class Company extends Model {

    protected $casts = [
        'open_date' => 'date',
    ];

}
```

Now I have the following Livewire component with a simple date input on it (it uses model data binding):

```php
class UserCompaniesDetail extends Component
{
    public Company $company;
    public $users = [];

    protected $rules = [
        'company.name' => ['required', 'max:255', 'string'],
        'company.open_date' => ['required', 'date', 'date'],
    ];
}
```

 but, whenever I try to update the date on a date HTML input, it breaks with a similar message:

```
The specified value "2021-06-18T00:00:00.000000Z" does not conform to the required format, "yyyy-MM-dd".
```
I noticed that Livewire always get the date cast from the Carbon instance:

![image](https://user-images.githubusercontent.com/11933789/123338431-35e78580-d51f-11eb-90c5-80eae22146d6.png)

I also tried to change the date using some hooks like *hydrateCompany, dehydrateCompany, updatingCompanyOpenDate*, but without success, it is always sent to the frontend in the *0000-00-00T00:00:00.000000Z* format.

I know that Livewire works when specifying a casting format directly on the model:

```php
class Company extends Model {

    protected $casts = [
        'open_date' => 'date:Y-m-d',
    ];

}
```

But it would break my entire application, as the code takes into account that the dates are cast to Carbon instances.

So, my question is:

**Is there a way to change the date before it is sent to the Livewire frontend component without the need to change the casting rules on my Model?**

Thank you very much for your help.

**Comment by TiagoSilvaPereira** [comment]
I'm considering separating the date properties on my component, then set it on my model after the update, but I would love to know if there is a solution using only the model data binding, without the need for separate properties.

**Comment by ph7jack** [comment]
You can use a date picker component like https://livewire-wireui.com/docs/datetime-picker

The Livewire casts the date in render if are using this date as carbon

**Comment by encenta** [comment]
Appreciate this is an old post but I'm having the same issue. I've scoured through the docs and wrote some logic similar to the OP but still can't resolve this. Is there really no way to format the date on frontend inputs without separating the properties from the model?

**Comment by stefanocurnis** [comment]
There are old discussion, closed issues and unmerged pull requests about this problem.
Livewire is still unusable as of today for this kind of thing. You basically can't model an input box to Carbon property or an integer property, a money property, ... Validation is useless if you can't stop the update. This forces us to use arrays everywhere instead of models.

[https://github.com/livewire/livewire/discussions/4610](url)

**Comment by joshhanley** [comment]
You can use an updating hook to convert a date string to a Carbon instance, see my comment on this thread for details https://github.com/livewire/livewire/discussions/3370#discussioncomment-1023847

Hope this helps!

**Comment by MarienMupenda** [comment]
```php
class Company extends Model {

    protected $casts = [
        //'open_date' => 'date:Y-m-d',
    ];

    public function getOpenDateAttribute()
    {
        return Carbon::parse($this->open_date);
    }
```

@TiagoSilvaPereira you can disable casting and make an accessor for that operation.

**Comment by mrl22** [comment, marked as answer]
Sorry for bringing up an old thread. I had this same issue where the date was set to the day before.

I fixed this by changing my cast in my model from:

```
protected $casts = [
        'date' => 'date'
];
```

to:

```
protected $casts = [
        'date' => 'date:Y-m-d'
];
```

**Reply by akamasAT** [reply]
after putting this topic on the long bench and now coming back to recheck and seeing this - boom - this solves it for me!

**Comment by eliasmaia** [comment]
May of 2025 and this problem still happens.

**Reply by akamasAT** [reply]
use the solution above - it has no drawbacks.

**Reply by eliasmaia** [reply]
Hi @akamasAT , thanks for replying.

I replicated those steps above but did not succeed. Do you have another suggestion  ? could you elaborate on that if possible?

- Tried doing the solution proposed https://github.com/livewire/livewire/discussions/3184#discussioncomment-7126193 but I get this same error message below
- Also tried @MarienMupenda's https://github.com/livewire/livewire/discussions/3184#discussioncomment-4475266 suggestion. It doesn't work either

![image](https://github.com/user-attachments/assets/ae07388f-da50-4a09-9541-c4c5cd66edf6)

**Reply by eliasmaia** [reply]
- This is my **Date component**, which is a component to render a input of **type=date**. 


```
<?php

namespace App\View\Components\Form\Input;

use Illuminate\View\Component;

class Date extends Component
{
    public string $id;
    public string $label;
    public string $model;
    public bool $required;
    public bool $readonly;

    public function __construct(
        string $id = '',
        string $label = '',
        string $model = '',
        bool $required = false,
        bool $readonly = false,
    ) {
        $this->id = $id;
        $this->label = $label;
        $this->model = $model;
        $this->required = $required;
        $this->readonly = $readonly;
    }

    public function render(): string
    {
        return <<<'blade'
            <x-form.input {{ $attributes->only('class') }} for="{{ $id }}" label="{{ $label }}" required="{{ $required }}">
                @if ($readonly)
                    <input {{ $attributes->except('class') }} type="date" id="{{ $id }}" wire:model="{{ $model }}" class="ml-2.5 py-1 text-gray-900 border-none p-0 w-full whitespace-nowrap" readonly disabled>
                @else
                    <input {{ $attributes->except('class') }} type="date" id="{{ $id }}" wire:model.live.debounce.500ms="{{ $model }}" class="bg-gray-50 border text-gray-900 text-sm rounded-lg focus:ring-blue-500 focus:border-blue-500 block w-full p-2.5 @error($model) border-red-600 @enderror">
                    <div>
                        @error($model) <span class="text-sm font-medium text-red-600">{{ $message }}</span> @enderror
                    </div>
                @endif
            </x-form.input>
        blade;
    }
}

```

- This is my **Livewire Form**, where I declare a few **date variables**:

```
<?php

namespace App\Livewire\Health\Forms;

use Carbon\Carbon;
use Livewire\Attributes\Validate;
use Livewire\Form;

class GynecologicalHistoryForm extends Form
{
    #[Validate('required', onUpdate: false)]
    public ?Carbon $mother_last_menstrual_period_date = null;
    #[Validate('required', onUpdate: false)]
    public ?Carbon $mother_last_pap_smear_test_date = null;
}
```

-  I actually create a property of **GynecologicalHistoryForm** in another Livewire component (**MedicalCheckup**), and handle the save/create/render of **GynecologicalHistoryForm** form inside this another Livewire component.

```
class MedicalCheckup extends Component
{
    public GynecologicalHistoryForm $gynecologicalHistoryForm;
    ...
   public function save() {...}
   public function render() {...}
}
```

- So in my blade view, this is how I declare the fields binded the dates declared:

```
<x-form.input.group title='Histórico Ginecológico'>
       <x-form.input.date label='Fecha de Última Menstruación ' id='mother_last_menstrual_period_date' model='gynecologicalHistoryForm.mother_last_menstrual_period_date' required :$readonly/>
       <x-form.input.date label='Última Prueba de Papanicolau' id='mother_last_pap_smear_test_date' model='gynecologicalHistoryForm.mother_last_pap_smear_test_date' required :$readonly/>
</x-form.input.group>
```


- I am declaring dates as Carbon. I tried before declaring them as DateTime, I experience the same bugs.
- When I declare them as strings, this does not happen.

Found this links of issues related to what I am facing, I also tried their suggestions and none of them worked: 
- https://github.com/livewire/livewire/discussions/8894
- https://laracasts.com/discuss/channels/livewire/bind-livewire-carbon-property-to-default-html-datepicker
- https://stackoverflow.com/questions/64064239/laravel-livewire-bind-model-carbon-attribute
- https://github.com/livewire/livewire/pull/1665
- https://laravel-livewire.com/docs/2.x/upgrading#casters
- https://github.com/livewire/livewire/discussions/3143
- https://github.com/livewire/livewire/discussions/3400

Hope to see a definite solution for this in the near future. If someone finds something else that might work for this problem, don't hesitate to @mention me here or reach me out. Thanks in advance

**Reply by akamasAT** [reply]
the original question is related to problems with casts of a model, but your problem has the same "root-problem". An HTML Input cannot handle a Carbon Object by default. If you echo a Carbon object, it includes the timestamp, a date input cannot handle this. your properties mother_last_menstrual_period_date and mother_last_pap_smear_test_date should be of type string and you need a setter method in your form that transforms your Carbon date object to string.

something like this
```
public funciton setDate(Carbon $date) {
    $this->mother_last_menstrual_period_date = $date->toDateString();
}
```

you can parse it back to carbon later if you need
